lib/elixir_sense/core/metadata_builder.ex

```diff
@@ -1663,6 +1668,98 @@ defmodule ElixirSense.Core.MetadataBuilder do
+        do: List.update_at(acc, index, &Map.put(&1, :type, type)),
+        else: acc
+    end)
+  end
+
+  # A guard expression can be in either these form:
+  #        :and                          :or
+  #      /     \            or          /   \              or      guard_expr
+  # guard_expr  guard_expr        guard_expr  guard_expr
+  #
+  # type information from :and subtrees are mergeable
+  # type information from :or subtrees are discarded
+  defp type_information_from_guards({:and, _, [guard_l, guard_r]}, state) do
+    left = type_information_from_guards(guard_l, state)
+    right = type_information_from_guards(guard_r, state)
+
+    Keyword.merge(left, right, fn _k, v1, v2 ->
+      case {v1, v2} do
+        # func my_func(x) when is_map_key(x, :a) and is_map_key(x, :b)
+        {{:map, fields1, _}, {:map, fields2, _}} ->
+          {:map, Enum.uniq_by(fields1 ++ fields2, &elem(&1, 0)), nil}
+
+        # In case we can't merge, just pick one
+        _ ->
+          v1
+      end
+    end)
+  end
+
+  defp type_information_from_guards({:or, _, [_guard_l, _guard_r]}, _state), do: []
+
+  defp type_information_from_guards(guard_ast, state) do
+    {_, acc} =
+      Macro.prewalk(guard_ast, [], fn
+        # Standalone variable: func my_func(x) when x
+        {var, _, nil} = node, acc ->
+          {node, [{var, :boolean} | acc]}
+
+        {guard_predicate, _, params} = node, acc ->
+          if type = guard_predicate_type(guard_predicate, params, state) do
+            [{var, _, nil} | _] = params
+            # If we found the predicate type, we can prematurely exit traversing the subtree
+            {[], [{var, type} | acc]}
+          else
+            {node, acc}
+          end
+
+        node, acc ->
+          {node, acc}
+      end)
+
+    acc
+  end
+
+  defp guard_predicate_type(p, _, _)
+       when p in [:is_number, :is_float, :is_integer, :round, :trunc, :div, :rem, :abs],
+       do: :number
+
+  defp guard_predicate_type(p, _, _) when p in [:is_binary, :binary_part], do: :binary
+
+  defp guard_predicate_type(p, _, _) when p in [:is_bitstring, :bit_size, :byte_size],
+    do: :bitstring
+
+  defp guard_predicate_type(p, _, _) when p in [:is_list, :hd, :tl, :length], do: :list
+  defp guard_predicate_type(p, _, _) when p in [:is_tuple, :tuple_size, :elem], do: :tuple
+  defp guard_predicate_type(:is_map, _, _), do: {:map, [], nil}
+
+  defp guard_predicate_type(:is_map_key, [_, key], state) do
+    case get_binding_type(state, key) do
+      {:atom, key} -> {:map, [{key, nil}], nil}
+      nil when is_binary(key) -> {:map, [{key, nil}], nil}
+      _ -> {:map, [], nil}
+    end
+  end
+
+  defp guard_predicate_type(:is_atom, _, _), do: :atom
+  defp guard_predicate_type(:is_boolean, _, _), do: :boolean
+
+  defp guard_predicate_type(:is_struct, [_, {:__aliases__, _, list}], state) do
+    {:struct, [], {:atom, expand_alias(state, list)}, nil}
+  end
+
+  defp guard_predicate_type(:is_struct, _, _), do: :struct
```

Comment:
```suggestion
  defp guard_predicate_type(:is_struct, _, _), do: {:struct, [], nil, nil}
```

Reply:
Fixed in ea32947181778082dcb8e9595fd96ceaa7351afc